File: solutions/ruby/kindergarten-garden/1/kindergarten_garden.rb
class Garden
  DEFAULT_STUDENTS = %w[Alice Bob Charlie David Eve Fred Ginny Harriet Ileana Joseph Kincaid Larry].freeze
  PLANTS = {
    'G' => :grass,
    'C' => :clover,
    'R' => :radishes,
    'V' => :violets
  }.freeze

  def initialize(diagram, students = DEFAULT_STUDENTS)
    @rows = diagram.split("\n")
    @students = students.sort
    @students.each_with_index do |student, index|
      define_singleton_method(student.downcase) do
        plants_for(index)
      end
    end
  end

  private

  def plants_for(index)
    start_pos = index * 2
    chars = []
    @rows.each do |row|
      chars << row[start_pos]
      chars << row[start_pos + 1]
    end
    chars.compact.map { |c| PLANTS[c] }
  end
end

File: solutions/python/kindergarten-garden/1/kindergarten_garden.py
class Garden:
    PLANTS = {
        'V': 'Violets',
        'R': 'Radishes',
        'C': 'Clover',
        'G': 'Grass'
    }
    
    DEFAULT_STUDENTS = [
        "Alice", "Bob", "Charlie", "David",
        "Eve", "Fred", "Ginny", "Harriet",
        "Ileana", "Joseph", "Kincaid", "Larry"
    ]

    def __init__(self, diagram, students=None):
        if students is None:
            students = self.DEFAULT_STUDENTS
        self.students = sorted(students)
        self.rows = diagram.split('\n')

    def plants(self, student):
        index = self.students.index(student) * 2
        res = []
        for row in self.rows:
            res.append(self.PLANTS[row[index]])
            res.append(self.PLANTS[row[index+1]])
        return res
